File: code/jvm/repository/src/testFixtures/java/pt/isel/repositories/mem/TransactionManagerInMem.java
package pt.isel.repositories.mem;

import pt.isel.repositories.Transaction;
import pt.isel.repositories.TransactionManager;
import pt.isel.repositories.channels.ChannelMemberRepository;
import pt.isel.repositories.channels.ChannelRepository;
import pt.isel.repositories.invitations.InvitationRepository;
import pt.isel.repositories.messages.MessageRepository;
import pt.isel.repositories.security.TokenBlacklistRepository;
import pt.isel.repositories.users.UserRepository;

import java.util.function.Function;

public class TransactionManagerInMem implements TransactionManager {
    private final UserRepository repoUsers = new UserRepositoryInMem();
    private final ChannelRepository repoChannels = new ChannelRepositoryInMem();
    private final MessageRepository repoMessages = new MessageRepositoryInMem();
    private final ChannelMemberRepository repoMemberships = new ChannelMemberRepositoryInMem();
    private final InvitationRepository repoInvitations = new InvitationRepositoryInMem();
    private final TokenBlacklistRepository repoTokenBlacklist = new TokenBlacklistRepositoryInMem();

    @Override
    public <R> R run(Function<Transaction, R> block) {
        return block.apply(new Transaction() {
            @Override
            public UserRepository repoUsers() {
                return repoUsers;
            }

            @Override
            public ChannelRepository repoChannels() {
                return repoChannels;
            }

            @Override
            public MessageRepository repoMessages() {
                return repoMessages;
            }

            @Override
            public ChannelMemberRepository repoMemberships() {
                return repoMemberships;
            }

            @Override
            public InvitationRepository repoInvitations() {
                return repoInvitations;
            }

            @Override
            public TokenBlacklistRepository repoTokenBlacklist() {
                return repoTokenBlacklist;
            }

            @Override
            public void rollback() { /* No-op in memory */ }
        });
    }
}

File: code/jvm/repository/src/test/kotlin/pt/isel/repositories/mem/TransactionManagerInMem.kt
package pt.isel.repositories.mem

import pt.isel.repositories.Transaction
import pt.isel.repositories.TransactionManager

class TransactionManagerInMem : TransactionManager {
    private val repoUsers = UserRepositoryInMem()
    private val repoChannels = ChannelRepositoryInMem()
    private val repoMessages = MessageRepositoryInMem()
    private val repoMemberships = ChannelMemberRepositoryInMem()
    private val repoInvitations = InvitationRepositoryInMem()
    private val repoTokenBlacklist = TokenBlacklistRepositoryInMem()

    override fun <R> run(block: Transaction.() -> R): R =
        block(
            TransactionInMem(
                repoUsers,
                repoChannels,
                repoMessages,
                repoMemberships,
                repoInvitations,
                repoTokenBlacklist,
            ),
        )
}
